stripe/src/main/java/com/stripe/android/net/PollingNetworkHandler.java

```diff
@@ -0,0 +1,187 @@
+        mPublishableKey = publishableKey;
+        mIsInSingleThreadMode = sourceRetriever != null;
+        mSourceRetriever = sourceRetriever == null
+                ? new SourceRetriever() {
+                    @Override
+                    public Source retrieveSource(
+                            @NonNull String sourceId,
+                            @NonNull String clientSecret,
+                            @NonNull String publishableKey) throws StripeException {
+                        return StripeApiHandler.retrieveSource(
+                                sourceId,
+                                clientSecret,
+                                publishableKey);
+                    }
+                }
+                : sourceRetriever;
+
+        mTimeoutMs = timeOutMs == null ? 10000L : timeOutMs.longValue();
+
+        mUiHandler = new Handler(Looper.getMainLooper()) {
+            int delayMs = INITIAL_DELAY_MS;
+            boolean terminated = false;
+
+            @Override
+            public void handleMessage(Message msg) {
+                super.handleMessage(msg);
+                if (terminated) {
+                    return;
+                }
+
+                switch (msg.what) {
+                    case SUCCESS:
+                        callback.onSuccess();
+                        break;
+                    case PENDING:
+                        delayMs *= POLLING_MULTIPLIER;
+                        callback.onRetry(delayMs);
+                        mNetworkHandler.sendEmptyMessage(delayMs);
+                        break;
+                    case FAILURE:
+                        terminated = true;
+                        callback.onError(new PollingFailedException(AUTHORIZATION_FAILED, false));
+                        break;
+                    case EXPIRED:
+                        terminated = true;
+                        callback.onError(new PollingFailedException(POLLING_EXPIRED, true));
+                        break;
+                    case ERROR:
+                        terminated = true;
+                        callback.onError((StripeException) msg.obj);
+                        break;
+                    default:
+                        break;
+                }
+            }
+        };
+
+        HandlerThread handlerThread = null;
+        if (!mIsInSingleThreadMode) {
+            handlerThread = new HandlerThread("Network Handler Thread");
+            handlerThread.start();
+        }
+
+        final Looper looper = mIsInSingleThreadMode
+                ? Looper.getMainLooper()
+                : handlerThread.getLooper();
+        mNetworkHandler = new Handler(looper) {
+            boolean terminated = false;
+
+            @Override
+            public void handleMessage(Message msg) {
+                super.handleMessage(msg);
+                if (terminated) {
+                    return;
+                }
+
+                if (msg.what >= 0) {
```

Comment:
This is a bit of a hack to communicate the number of milliseconds to delay the next call to the Runnable task, but I have no other information to communicate to the network thread, and it is very compact.